test/tests_for_sdkmr/EE2Runjob_test.py

```diff
@@ -267,6 +273,52 @@ def _create_reqs_dict(
     return ret
 
 
+def _setup_test_preflight_batch_logic():
+    mocks = _set_up_mocks(_USER, _TOKEN)
+    sdkmr = mocks[SDKMethodRunner]
+    _set_up_common_return_values(mocks)
+    rj = EE2RunJob(sdkmr)
+    rj._check_workspace_permissions_list = MagicMock()
+    rj._handle_job_requirements = MagicMock()
+    rj._check_job_arguments = MagicMock()
+    rj._run = MagicMock()
+    params = {
+        "method": _METHOD,
+        "app_id": _APP,
+        "source_ws_objects": [_WS_REF_1, _WS_REF_2],
+        "wsid": "123",
+    }
+    jrr = mocks[JobRequirementsResolver]
+    jrr.normalize_job_reqs.return_value = {}
+    jrr.get_requirements_type.return_value = RequirementsType.STANDARD
+    return rj, params
+
+
+def test_preflight_logics():
+    rj, params = _setup_test_preflight_batch_logic()
+    # Check regular preflight
+    rj._preflight([params])
+    rj._check_workspace_permissions_list.assert_called_once_with([params.get("wsid")])
+    rj._handle_job_requirements.assert_called_once_with(
+        as_admin=False, concierge_params=None, params=[params]
+    )
+    rj._check_job_arguments.assert_called_once_with([params])
+
+    # Check Batch preflight
+    rj2, params2 = _setup_test_preflight_batch_logic()
+    params2["wsid"] = "1234"
+    rj2._preflight_batch([params, params2])
+    rj2._check_workspace_permissions_list.assert_called_once_with(
+        [params.get("wsid"), params2.get("wsid")]
+    )
+    rj2._handle_job_requirements.assert_called_once_with(
+        as_admin=False, params=[params, params2]
+    )
+    rj2._check_job_arguments.assert_called_once_with(
+        [params, params2], has_parent_job=True
```

Comment:
Darn, and I thought I was finally writing better tests by using assert_called_once_with to make sure that the logic was being called, and if something in that function was deleted, this test would break. What would you recommend instead?

Reply:
The thing is I *should* be able to rename / delete / modify internal methods / variables without having the tests break as long as it doesn't change the external behavior. The external behavior is what we care about, not the implementation details. For EE2RunJob, I don't care if `_preflight` is called as long as the job requirements are saved in mongo and passed to Condor.

As such, tests should ideally always be against the public API of the code unit and any mocks that need to be provided to the code unit for it to run. My general test rules go like this:

1) Test one method or function call per test (except for trivial getters, I'll usually test them all in one test)
1a) Unless the method changes class state - then you may need to check the state as well.
2) Create mocks for any other code units that the code unit under test requires
3) Set up the expected returns for the mocks (these should be public methods that you expect the code unit under test to call)
4) Pass the mocks into the constructor / method / function as necessary.
5) assert that the return value of the method, if any, is correct.
6) assert that calls to the mocks (again, public API) are correct.

That's what the tests in this module do, so given there's no new behavior or new methods in this PR, I think you should just have to adjust the tests for the minor changes to workspace checking, although I haven't check that in detail.

Reply:
Actually you may have added some more error checking to `run_one_job` now I think about it.